pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,244 @@
+        """
+        This function is used to compute the effective iteration.
+        As DataLoader can perform ``prefecthing`` or training can fail while processing a batch,
+        the current iteration needs to be computed using the ``num_batches_processed`` processed information.
+        """
+        if num_batches_processed is not None:
+            current_iteration = num_batches_processed
+        else:
+            current_iteration = self._current_iteration
+
+        if self._dataloader_batch_size:
+            current_iteration *= self._dataloader_batch_size
+
+        return current_iteration
+
+    def state_dict(self, num_batches_processed: Optional[int] = None) -> Dict[int, Dict[str, int]]:
+        """ Returns the state of the sampler in the current worker. The worker id indexes the state dict."""
+        return {self.worker_id: {"current_iteration": self._compute_current_iteration(num_batches_processed)}}
+
+    def load_state_dict(self, state_dict: Dict[int, Any], workers_initialized: bool = False) -> None:
+        """
+        Loads the saved state for the wrapped sampler.
+        If the ``state_dict`` contains multiple states, it means there were multiple workers.
+        The state will be cached and fully reloaded (fast-forward) the first time :meth:`__iter__` is called.
+        """
+        # as workers aren't available, the ``state_dict``` is cached until workers are made available.
+        if len(state_dict) > 1 and not workers_initialized:
+            self._cached_state_dict = deepcopy(state_dict)
+            self.restarting = self._cached_state_dict[self.worker_id]["current_iteration"] > 0
+            return
+        self._current_iteration = state_dict[self.worker_id]["current_iteration"]
+        self.restarting = self._current_iteration > 0
+
+
+class CaptureIterableDataset(IterableDataset):
+    """
+    The ``CaptureIterableDataset`` is used to wrap an :class:`torch.utils.data.IterableDataset`.
+    On ``__iter__`` function call   the ``CaptureIterableDataset`` will wrap the wrapped dataset
+        generators into ``FastForwardSampler`` to keep track of progress.
+    On ``__next__`` function call, the ``CaptureIterableDataset`` will return a dictionary containing
+        user data and metadata containing the ``FastForwardSampler`` samplers state_dict.
+    """
+
+    def __init__(self, dataset: IterableDataset, initial_seed: Optional[int] = None):
+        super().__init__()
+        self.dataset = dataset
+        self.state_dict: Optional[Dict[int, Any]] = None
+        self.initial_seed = initial_seed
+        self.samplers: Optional[Dict[str, FastForwardSampler]] = None
+
+    @property
+    def sampler(self) -> Sampler:
+        return self.dataset.sampler
+
+    def load_state_dict(self, state_dict: Dict[int, Any]) -> None:
+        self.state_dict = deepcopy(state_dict)
+
+    def _wrap_generator_samplers(self) -> None:
+        if self.samplers is None:
```

Comment:
Avoid the indentation

```suggestion
        if self.samplers is not None:
            return
```

Reply:
Done !